File: packages/go/receipts.go
package rdk

import (
	"context"
	"encoding/binary"
	"fmt"

	pqc "github.com/qorechain/qorechain-pqc/go"
)

// RECEIPT_ALGORITHM is the post-quantum algorithm the anchor signature uses.
const ReceiptAlgorithm = "ML-DSA-87"

// ReceiptVersion is the current receipt schema version.
const ReceiptVersion = 1

// SettlementReceipt is a portable record that a rollup's settlement batch was
// anchored to the QoreChain Main Chain under a post-quantum (ML-DSA-87 /
// Dilithium-5) signature.
//
// A receipt is a claim, not evidence — every field in it is supplied by whoever
// hands it to you. VerifySettlementReceipt therefore re-reads the claim from
// live chain state (the rollup's layer, the batch's state root, the anchor, and
// the creator's registered key) and only then reports it valid. Checking the
// signature alone, against a key you were handed, proves that someone signed
// those bytes — not that QoreChain anchored anything.
//
// Canonical anchor message (matches the chain's anchorSignBytes):
//
//	layer_id || layer_height(8-byte big-endian) || state_root || validator_set_hash
type SettlementReceipt struct {
	Version    int
	RollupID   string
	LayerID    string
	BatchIndex uint64
	// Creator is the layer creator — the registered signer of the anchor's PQC
	// signature.
	Creator   string
	Algorithm string
	// StateRoot is the anchored state root (hex).
	StateRoot        string
	LayerHeight      uint64
	ValidatorSetHash string
	MainChainHeight  uint64
	AnchoredAt       int64
	// PqcSignature is the Dilithium-5 anchor signature (hex).
	PqcSignature string
	// BatchStateRoot is the state root read from the settlement batch (hex) when
	// the receipt was built. Informational only — verification compares the
	// receipt against the chain's batch, never against this copy of it.
	BatchStateRoot string
}

// ReceiptVerificationMode says how a receipt was checked.
type ReceiptVerificationMode string

const (
	// ReceiptModeChain means every field was reproduced from live chain state.
	// Only this mode can yield Valid: true.
	ReceiptModeChain ReceiptVerificationMode = "chain"
	// ReceiptModeSignatureOnly means only the ML-DSA-87 signature was checked,
	// against a key the caller supplied. That proves whoever holds that key
	// signed these bytes; it does NOT prove the anchor exists on QoreChain, so
	// the result is always Valid: false.
	ReceiptModeSignatureOnly ReceiptVerificationMode = "signature-only"
)

// ReceiptChecks records the individual checks performed when verifying a
// receipt. Every check is answered from chain state, never from the receipt
// itself.
type ReceiptChecks struct {
	// RollupLayerBinding is true when the chain says the receipt's rollup
	// belongs to the receipt's layer.
	RollupLayerBinding bool
	// BatchStateRoot is true when the chain's batch carries the receipt's state
	// root.
	BatchStateRoot bool
	// AnchorOnChain is true when an anchor with this state root, height,
	// validator-set hash and signature exists on chain.
	AnchorOnChain bool
	// CreatorAuthority is true when the receipt names the chain's rollup creator
	// and that creator has a registered post-quantum key (resolved from chain,
	// not taken from the receipt).
	CreatorAuthority bool
	// PqcSignature is true when the Dilithium-5 signature over the canonical
	// message verified.
	PqcSignature bool
}

// ReceiptVerification is the outcome of verifying a receipt.
type ReceiptVerification struct {
	// Valid is true only when the receipt was reproduced from live chain state.
	Valid bool
	// Mode records which check was actually performed.
	Mode   ReceiptVerificationMode
	Checks ReceiptChecks
	Reason string
}

// AnchorSignBytes reconstructs the canonical message the chain signs for a state
// anchor:
//
//	layer_id || layer_height(8B BE) || state_root || validator_set_hash
//
// stateRootHex and vshHex are hex strings (with or without a 0x prefix) and
// layerHeight is a uint64. Invalid hex inputs are treated as empty byte runs, so
// callers that have validated their inputs get an exact canonical encoding.
func AnchorSignBytes(layerID string, layerHeight uint64, stateRootHex, vshHex string) []byte {
	height := make([]byte, 8)
	binary.BigEndian.PutUint64(height, layerHeight)
	stateRoot, _ := HexToBytes(stateRootHex)
	vsh, _ := HexToBytes(vshHex)
	out := make([]byte, 0, len(layerID)+len(height)+len(stateRoot)+len(vsh))
	out = append(out, []byte(layerID)...)
	out = append(out, height...)
	out = append(out, stateRoot...)
	out = append(out, vsh...)
	return out
}

// BuildSettlementReceipt builds a settlement receipt for rollupID's batch
// batchIndex: it resolves the rollup's layer, reads the batch's state root, and
// finds the state anchor that commits that root to the Main Chain. It errors if
// the rollup has no layer, the batch is missing, or no anchor covers the batch's
// state root yet.
func BuildSettlementReceipt(ctx context.Context, client *RdkClient, rollupID string, batchIndex uint64) (SettlementReceipt, error) {
	rollup, err := client.Rest.GetRollup(ctx, rollupID)
	if err != nil {
		return SettlementReceipt{}, err
	}
	if rollup.LayerID == "" {
		return SettlementReceipt{}, fmt.Errorf("rollup %q has no layer_id — it is not anchored to a multilayer layer", rollupID)
	}
	batch, err := client.Rest.GetBatch(ctx, rollupID, batchIndex)
	if err != nil {
		return SettlementReceipt{}, err
	}
	batchStateRootHex := hexWireBytes(batch.StateRoot)

	anchors, err := client.Rest.GetAnchors(ctx, rollup.LayerID)
	if err != nil {
		return SettlementReceipt{}, err
	}
	var covering *AnchorView
	for i := range anchors {
		if anchors[i].StateRoot != "" && anchors[i].StateRoot == batchStateRootHex {
			covering = &anchors[i]
			break
		}
	}
	anchor := covering
	if anchor == nil {
		latest, lerr := client.Rest.GetLatestAnchor(ctx, rollup.LayerID)
		if lerr != nil {
			return SettlementReceipt{}, lerr
		}
		anchor = &latest
	}
	if anchor.StateRoot == "" {
		return SettlementReceipt{}, fmt.Errorf("no state anchor found for layer %q", rollup.LayerID)
	}
	if covering == nil {
		return SettlementReceipt{}, fmt.Errorf(
			"no anchor commits batch %d's state root yet (latest anchored height %d); the batch may not be anchored to the Main Chain",
			batchIndex, anchor.LayerHeight)
	}
	return SettlementReceipt{
		Version:          ReceiptVersion,
		RollupID:         rollupID,
		LayerID:          rollup.LayerID,
		BatchIndex:       batchIndex,
		Creator:          rollup.Creator,
		Algorithm:        ReceiptAlgorithm,
		StateRoot:        anchor.StateRoot,
		LayerHeight:      anchor.LayerHeight,
		ValidatorSetHash: anchor.ValidatorSetHash,
		MainChainHeight:  anchor.MainChainHeight,
		AnchoredAt:       anchor.AnchoredAt,
		PqcSignature:     anchor.PqcSignature,
		BatchStateRoot:   batchStateRootHex,
	}, nil
}

// signatureOver checks the receipt's Dilithium-5 anchor signature against
// publicKeyHex over the canonical anchor message.
func signatureOver(receipt SettlementReceipt, publicKeyHex string) bool {
	if publicKeyHex == "" || receipt.PqcSignature == "" {
		return false
	}
	pub, err := HexToBytes(publicKeyHex)
	if err != nil {
		return false
	}
	sig, err := HexToBytes(receipt.PqcSignature)
	if err != nil {
		return false
	}
	message := AnchorSignBytes(receipt.LayerID, receipt.LayerHeight, receipt.StateRoot, receipt.ValidatorSetHash)
	return pqc.MLDSA87.Verify(pub, message, sig)
}

// VerifySettlementReceipt verifies a settlement receipt against live chain
// state.
//
// A receipt is a claim, not evidence: every field in it is supplied by whoever
// hands it to you. Verification therefore re-reads the claim from the chain —
// the rollup's layer, the batch's state root, the anchor itself, and the signing
// key — and reports Valid: true (with Mode ReceiptModeChain) only when the
// receipt reproduces what the chain actually holds. Pass a non-nil client for
// this.
//
// Without a client, creatorPublicKeyHex (the layer creator's ML-DSA-87 public
// key in hex, obtained out-of-band) still lets you check the signature, but that
// is Mode ReceiptModeSignatureOnly and never Valid: a signature proves someone
// signed these bytes, not that the anchor exists on QoreChain. Verification is
// NOT possible offline. creatorPublicKeyHex is ignored when client is supplied —
// the key is always resolved from chain state in that case.
func VerifySettlementReceipt(ctx context.Context, receipt SettlementReceipt, creatorPublicKeyHex string, client *RdkClient) ReceiptVerification {
	checks := ReceiptChecks{}

	if client == nil {
		if creatorPublicKeyHex == "" {
			return ReceiptVerification{
				Valid:  false,
				Mode:   ReceiptModeSignatureOnly,
				Checks: checks,
				Reason: "no client supplied — pass a client to verify against chain state; a receipt on its own proves nothing",
			}
		}
		checks.PqcSignature = signatureOver(receipt, creatorPublicKeyHex)
		reason := "signature did not verify against the supplied key"
		if checks.PqcSignature {
			reason = "signature is valid for the supplied key, but nothing was checked against the chain — pass a client to verify settlement"
		}
		return ReceiptVerification{Valid: false, Mode: ReceiptModeSignatureOnly, Checks: checks, Reason: reason}
	}

	if ctx == nil {
		ctx = context.Background()
	}
	rest := client.Rest
	fail := func(format string, args ...any) ReceiptVerification {
		return ReceiptVerification{
			Valid:  false,
			Mode:   ReceiptModeChain,
			Checks: checks,
			Reason: fmt.Sprintf(format, args...),
		}
	}

	// 1. The rollup must exist on chain and belong to the layer the receipt names.
	rollup, err := rest.GetRollup(ctx, receipt.RollupID)
	if err != nil {
		return fail("rollup %q not found on chain: %v", receipt.RollupID, err)
	}
	checks.RollupLayerBinding = rollup.LayerID != "" && rollup.LayerID == receipt.LayerID
	if !checks.RollupLayerBinding {
		onChainLayer := rollup.LayerID
		if onChainLayer == "" {
			onChainLayer = "(none)"
		}
		return fail("rollup %q is anchored to layer %q, not %q", receipt.RollupID, onChainLayer, receipt.LayerID)
	}

	// 2. The chain's batch must carry the receipt's state root (never the
	//    receipt's own copy of it).
	batch, err := rest.GetBatch(ctx, receipt.RollupID, receipt.BatchIndex)
	if err != nil {
		return fail("batch %d not found on chain: %v", receipt.BatchIndex, err)
	}
	onChainRoot := hexWireBytes(batch.StateRoot)
	checks.BatchStateRoot = onChainRoot != "" && onChainRoot == receipt.StateRoot
	if !checks.BatchStateRoot {
		return fail("the chain's batch does not carry the receipt's state root")
	}

	// 3. An anchor matching this receipt must actually exist on chain.
	anchors, err := rest.GetAnchors(ctx, receipt.LayerID)
	if err != nil {
		return fail("could not read anchors for layer %q: %v", receipt.LayerID, err)
	}
	for _, a := range anchors {
		if a.StateRoot == receipt.StateRoot &&
			a.LayerHeight == receipt.LayerHeight &&
			a.ValidatorSetHash == receipt.ValidatorSetHash &&
			a.PqcSignature == receipt.PqcSignature {
			checks.AnchorOnChain = true
			break
		}
	}
	if !checks.AnchorOnChain {
		return fail("no anchor on chain matches this receipt — it is fabricated or superseded")
	}

	// 4. The signing key is resolved from the chain, never taken from the receipt.
	account, err := rest.GetPqcAccount(ctx, rollup.Creator)
	if err != nil {
		return fail("could not resolve the creator's post-quantum key: %v", err)
	}
	if receipt.Creator != rollup.Creator {
		return fail("receipt names creator %q but the chain says the rollup's creator is %q", receipt.Creator, rollup.Creator)
	}
	checks.CreatorAuthority = account.PublicKey != ""
	if !checks.CreatorAuthority {
		return fail("no post-quantum key is registered for creator %s", rollup.Creator)
	}

	// 5. Finally the post-quantum signature over the canonical anchor message.
	checks.PqcSignature = signatureOver(receipt, account.PublicKey)
	if !checks.PqcSignature {
		return fail("Dilithium-5 anchor signature did not verify against the creator's registered key")
	}
	return ReceiptVerification{Valid: true, Mode: ReceiptModeChain, Checks: checks}
}

File: packages/rust/src/receipts.rs
//! Quantum-Safe Settlement Receipts.
//!
//! A settlement receipt is a portable record that a rollup's settlement batch
//! was anchored to the QoreChain Main Chain under a post-quantum
//! (ML-DSA-87 / Dilithium-5) signature.
//!
//! A receipt is a **claim, not evidence** — every field in it is supplied by
//! whoever hands it to you. [`verify_settlement_receipt`] therefore re-reads the
//! claim from live chain state (the rollup's layer, the batch's state root, the
//! anchor, and the creator's registered key) and only then reports it valid.
//! Checking the signature alone, against a key you were handed, proves that
//! someone signed those bytes — not that QoreChain anchored anything. There is
//! no offline verification of a settlement receipt.
//!
//! Canonical anchor message (matches the chain's `anchorSignBytes`):
//! `layer_id || layer_height(8-byte big-endian) || state_root || validator_set_hash`.

use serde::{Deserialize, Serialize};

use crate::client::facade::RdkClient;
use crate::client::views::AnchorView;
use crate::utils::bytes::{bytes_to_hex, decode_wire_bytes, hex_to_bytes};

/// The post-quantum algorithm the anchor signature uses.
pub const RECEIPT_ALGORITHM: &str = "ML-DSA-87";

/// Current receipt schema version.
pub const RECEIPT_VERSION: u32 = 1;

/// An error building or verifying a settlement receipt.
#[derive(Debug, Clone, thiserror::Error)]
pub enum ReceiptError {
    /// A REST read failed.
    #[error(transparent)]
    Rest(#[from] crate::client::rest::RestError),
    /// The rollup is not anchored to a multilayer layer.
    #[error("rollup \"{0}\" has no layer_id — it is not anchored to a multilayer layer")]
    NoLayer(String),
    /// No state anchor exists for the layer yet.
    #[error("no state anchor found for layer \"{0}\"")]
    NoAnchor(String),
    /// No anchor commits the batch's state root yet.
    #[error(
        "no anchor commits batch {batch_index}'s state root yet (latest anchored height {latest_height}); the batch may not be anchored to the Main Chain"
    )]
    NotAnchored {
        /// The batch index requested.
        batch_index: u64,
        /// The latest anchored layer height seen.
        latest_height: u64,
    },
}

/// A portable settlement receipt. Verify it with [`verify_settlement_receipt`],
/// which re-reads every field from live chain state — the receipt on its own is
/// a claim, not evidence.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SettlementReceipt {
    /// Receipt schema version.
    pub version: u32,
    /// The rollup id.
    pub rollup_id: String,
    /// The host layer id.
    pub layer_id: String,
    /// The settlement batch index.
    pub batch_index: u64,
    /// The layer creator — the registered signer of the anchor's PQC signature.
    pub creator: String,
    /// The post-quantum algorithm name.
    pub algorithm: String,
    /// The anchored state root (hex).
    pub state_root: String,
    /// The anchored layer height.
    pub layer_height: u64,
    /// The anchored validator-set hash (hex).
    pub validator_set_hash: String,
    /// The Main Chain height the anchor was committed at.
    pub main_chain_height: i64,
    /// Anchor unix timestamp (seconds).
    pub anchored_at: i64,
    /// The Dilithium-5 anchor signature (hex).
    pub pqc_signature: String,
    /// The state root read from the settlement batch (hex) when the receipt was
    /// built. **Informational only** — verification compares the receipt against
    /// the chain's batch, never against this copy of it.
    pub batch_state_root: String,
}

/// How a receipt was checked.
///
/// - [`ReceiptVerificationMode::Chain`] — every field was reproduced from live
///   chain state. Only this mode can yield `valid: true`.
/// - [`ReceiptVerificationMode::SignatureOnly`] — the ML-DSA-87 signature was
///   checked against a key the caller supplied. That proves whoever holds that
///   key signed these bytes; it does **not** prove the anchor exists on
///   QoreChain. Always `valid: false`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ReceiptVerificationMode {
    /// Verified against live chain state.
    Chain,
    /// Only the signature was checked, against a caller-supplied key.
    SignatureOnly,
}

/// The individual checks of a receipt verification.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReceiptChecks {
    /// The chain says `rollup_id` belongs to the receipt's `layer_id`.
    pub rollup_layer_binding: bool,
    /// The chain's batch carries the receipt's state root.
    pub batch_state_root: bool,
    /// An anchor with this state root, height and signature exists on chain.
    pub anchor_on_chain: bool,
    /// The signing key was resolved from chain, not taken from the receipt.
    pub creator_authority: bool,
    /// The Dilithium-5 signature over the canonical message verified.
    pub pqc_signature: bool,
}

/// The outcome of verifying a receipt.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReceiptVerification {
    /// True only when the receipt was reproduced from live chain state.
    pub valid: bool,
    /// Which check was actually performed.
    pub mode: ReceiptVerificationMode,
    /// The individual checks.
    pub checks: ReceiptChecks,
    /// A human-readable reason when invalid.
    pub reason: Option<String>,
}

/// Reconstruct the canonical message the chain signs for a state anchor:
/// `layer_id || layer_height(8B BE) || state_root || validator_set_hash`.
/// Inputs are taken in hex (`state_root_hex`, `vsh_hex`) and a numeric height.
pub fn anchor_sign_bytes(
    layer_id: &str,
    layer_height: u64,
    state_root_hex: &str,
    vsh_hex: &str,
) -> Vec<u8> {
    let state_root = hex_to_bytes(state_root_hex).unwrap_or_default();
    let vsh = hex_to_bytes(vsh_hex).unwrap_or_default();
    let mut out = Vec::with_capacity(layer_id.len() + 8 + state_root.len() + vsh.len());
    out.extend_from_slice(layer_id.as_bytes());
    out.extend_from_slice(&layer_height.to_be_bytes());
    out.extend_from_slice(&state_root);
    out.extend_from_slice(&vsh);
    out
}

fn receipt_from_parts(
    rollup_id: &str,
    layer_id: &str,
    batch_index: u64,
    creator: &str,
    anchor: &AnchorView,
    batch_state_root_hex: &str,
) -> SettlementReceipt {
    SettlementReceipt {
        version: RECEIPT_VERSION,
        rollup_id: rollup_id.to_string(),
        layer_id: layer_id.to_string(),
        batch_index,
        creator: creator.to_string(),
        algorithm: RECEIPT_ALGORITHM.to_string(),
        state_root: anchor.state_root.clone(),
        layer_height: anchor.layer_height,
        validator_set_hash: anchor.validator_set_hash.clone(),
        main_chain_height: anchor.main_chain_height,
        anchored_at: anchor.anchored_at,
        pqc_signature: anchor.pqc_signature.clone(),
        batch_state_root: batch_state_root_hex.to_string(),
    }
}

/// Build a settlement receipt for `rollup_id`'s batch `batch_index`: resolve the
/// rollup's layer, read the batch's state root, and find the state anchor that
/// commits that root to the Main Chain. Errors if the rollup has no layer, the
/// batch is missing, or no anchor covers the batch's state root yet.
pub fn build_settlement_receipt(
    client: &RdkClient,
    rollup_id: &str,
    batch_index: u64,
) -> Result<SettlementReceipt, ReceiptError> {
    let rollup = client.rest.get_rollup(rollup_id)?;
    if rollup.layer_id.is_empty() {
        return Err(ReceiptError::NoLayer(rollup_id.to_string()));
    }
    let batch = client.rest.get_batch(rollup_id, batch_index)?;
    let batch_state_root_hex = if batch.state_root.is_empty() {
        String::new()
    } else {
        bytes_to_hex(&decode_wire_bytes(&batch.state_root))
    };

    let anchors = client.rest.get_anchors(&rollup.layer_id)?;
    let covering = anchors
        .iter()
        .find(|a| !a.state_root.is_empty() && a.state_root == batch_state_root_hex)
        .cloned();
    let anchor = match covering.clone() {
        Some(a) => a,
        None => client.rest.get_latest_anchor(&rollup.layer_id)?,
    };
    if anchor.state_root.is_empty() {
        return Err(ReceiptError::NoAnchor(rollup.layer_id.clone()));
    }
    if covering.is_none() {
        return Err(ReceiptError::NotAnchored {
            batch_index,
            latest_height: anchor.layer_height,
        });
    }
    Ok(receipt_from_parts(
        rollup_id,
        &rollup.layer_id,
        batch_index,
        &rollup.creator,
        &anchor,
        &batch_state_root_hex,
    ))
}

/// Check the receipt's ML-DSA-87 anchor signature against `public_key_hex`.
fn signature_over(receipt: &SettlementReceipt, public_key_hex: &str) -> bool {
    if public_key_hex.is_empty() || receipt.pqc_signature.is_empty() {
        return false;
    }
    let (Ok(public_key), Ok(signature)) = (
        hex_to_bytes(public_key_hex),
        hex_to_bytes(&receipt.pqc_signature),
    ) else {
        return false;
    };
    let message = anchor_sign_bytes(
        &receipt.layer_id,
        receipt.layer_height,
        &receipt.state_root,
        &receipt.validator_set_hash,
    );
    qorechain_pqc::mldsa::ml_dsa_87::verify(&public_key, &message, &signature)
}

/// Verify a settlement receipt against live chain state.
///
/// A receipt is a *claim*, not evidence: every field in it is supplied by
/// whoever hands it to you. Verification therefore re-reads the claim from the
/// chain — the rollup's layer, the batch's state root, the anchor itself, and
/// the signing key — and reports `valid: true` only when the receipt reproduces
/// what the chain actually holds. Pass `client` for this.
///
/// Without a `client` you can still check the signature against a
/// `creator_public_key_hex` you obtained out of band, but that is
/// [`ReceiptVerificationMode::SignatureOnly`] and never `valid`: a signature
/// proves someone signed these bytes, not that the anchor exists. When `client`
/// is supplied, `creator_public_key_hex` is ignored — the key is resolved from
/// chain.
pub fn verify_settlement_receipt(
    receipt: &SettlementReceipt,
    creator_public_key_hex: Option<&str>,
    client: Option<&RdkClient>,
) -> ReceiptVerification {
    let mut checks = ReceiptChecks::default();

    let Some(client) = client else {
        let Some(public_key_hex) = creator_public_key_hex else {
            return ReceiptVerification {
                valid: false,
                mode: ReceiptVerificationMode::SignatureOnly,
                checks,
                reason: Some(
                    "no client supplied — pass a client to verify against chain state; a receipt on its own proves nothing"
                        .to_string(),
                ),
            };
        };
        checks.pqc_signature = signature_over(receipt, public_key_hex);
        let reason = if checks.pqc_signature {
            "signature is valid for the supplied key, but nothing was checked against the chain — pass a client to verify settlement"
        } else {
            "signature did not verify against the supplied key"
        };
        return ReceiptVerification {
            valid: false,
            mode: ReceiptVerificationMode::SignatureOnly,
            checks,
            reason: Some(reason.to_string()),
        };
    };

    let fail = |checks: ReceiptChecks, reason: String| ReceiptVerification {
        valid: false,
        mode: ReceiptVerificationMode::Chain,
        checks,
        reason: Some(reason),
    };

    // 1. The rollup must exist on chain and belong to the layer the receipt names.
    let rollup = match client.rest.get_rollup(&receipt.rollup_id) {
        Ok(r) => r,
        Err(e) => {
            return fail(
                checks,
                format!(
                    "rollup \"{}\" not found on chain: {e}",
                    receipt.rollup_id
                ),
            );
        }
    };
    checks.rollup_layer_binding =
        !rollup.layer_id.is_empty() && rollup.layer_id == receipt.layer_id;
    if !checks.rollup_layer_binding {
        let on_chain = if rollup.layer_id.is_empty() {
            "(none)"
        } else {
            rollup.layer_id.as_str()
        };
        return fail(
            checks,
            format!(
                "rollup \"{}\" is anchored to layer \"{on_chain}\", not \"{}\"",
                receipt.rollup_id, receipt.layer_id
            ),
        );
    }

    // 2. The chain's batch must carry the receipt's state root (not the receipt's own copy).
    let batch = match client.rest.get_batch(&receipt.rollup_id, receipt.batch_index) {
        Ok(b) => b,
        Err(e) => {
            return fail(
                checks,
                format!("batch {} not found on chain: {e}", receipt.batch_index),
            );
        }
    };
    let on_chain_root = if batch.state_root.is_empty() {
        String::new()
    } else {
        bytes_to_hex(&decode_wire_bytes(&batch.state_root))
    };
    checks.batch_state_root = !on_chain_root.is_empty() && on_chain_root == receipt.state_root;
    if !checks.batch_state_root {
        return fail(
            checks,
            "the chain's batch does not carry the receipt's state root".to_string(),
        );
    }

    // 3. An anchor matching this receipt must actually exist on chain.
    let anchors = match client.rest.get_anchors(&receipt.layer_id) {
        Ok(a) => a,
        Err(e) => {
            return fail(
                checks,
                format!(
                    "could not read anchors for layer \"{}\": {e}",
                    receipt.layer_id
                ),
            );
        }
    };
    checks.anchor_on_chain = anchors.iter().any(|a| {
        a.state_root == receipt.state_root
            && a.layer_height == receipt.layer_height
            && a.validator_set_hash == receipt.validator_set_hash
            && a.pqc_signature == receipt.pqc_signature
    });
    if !checks.anchor_on_chain {
        return fail(
            checks,
            "no anchor on chain matches this receipt — it is fabricated or superseded".to_string(),
        );
    }

    // 4. The signing key is resolved from the chain, never taken from the receipt.
    let public_key_hex = match client.rest.get_pqc_account(&rollup.creator) {
        Ok(account) => account.public_key,
        Err(e) => {
            return fail(
                checks,
                format!("could not resolve the creator's post-quantum key: {e}"),
            );
        }
    };
    if receipt.creator != rollup.creator {
        return fail(
            checks,
            format!(
                "receipt names creator \"{}\" but the chain says the rollup's creator is \"{}\"",
                receipt.creator, rollup.creator
            ),
        );
    }
    checks.creator_authority = !public_key_hex.is_empty();
    if !checks.creator_authority {
        return fail(
            checks,
            format!(
                "no post-quantum key is registered for creator {}",
                rollup.creator
            ),
        );
    }

    // 5. Finally the post-quantum signature over the canonical anchor message.
    checks.pqc_signature = signature_over(receipt, &public_key_hex);
    if !checks.pqc_signature {
        return fail(
            checks,
            "Dilithium-5 anchor signature did not verify against the creator's registered key"
                .to_string(),
        );
    }

    ReceiptVerification {
        valid: true,
        mode: ReceiptVerificationMode::Chain,
        checks,
        reason: None,
    }
}
